src/msgraph_core/tasks/page_iterator.py

```diff
@@ -0,0 +1,227 @@
+from kiota_abstractions.request_information import RequestInformation  # type: ignore
+from kiota_abstractions.serialization.parsable import Parsable  # type: ignore
+
+from msgraph_core.models import PageResult  # pylint: disable=no-name-in-module, import-error
+
+T = TypeVar('T', bound=Parsable)
+
+
+class PageIterator:
+    """
+This class is used to iterate over paged responses from a server.
+
+The PageIterator class provides methods to iterate over the items in the pages,
+fetch the next page, and convert a response to a page.
+
+Attributes:
+    request_adapter (HttpxRequestAdapter): The adapter used to send HTTP requests.
+    pause_index (int): The index at which to pause iteration.
+    headers (HeadersCollection): The headers to include in the HTTP requests.
+    request_options (list): The options for the HTTP requests.
+    current_page (PageResult): The current page of items.
+    object_type (str): The type of the items in the pages.
+    has_next (bool): Whether there are more pages to fetch.
+
+Methods:
+    __init__(response: Union[T, list, object], request_adapter: HttpxRequestAdapter,
+     constructor_callable: Optional[Callable] = None): Initializes a new instance of 
+     the PageIterator class.
+    """
+
+    def __init__(
+        self,
+        response: Union[T, list, object],
+        request_adapter: HttpxRequestAdapter,
+        constructor_callable: Optional[Callable] = None
+    ):
+        self.request_adapter = request_adapter
+
+        if isinstance(response, Parsable) and not constructor_callable:
+            constructor_callable = getattr(type(response), 'create_from_discriminator_value')
+        elif constructor_callable is None:
+            constructor_callable = PageResult.create_from_discriminator_value
+        self.pause_index = 0
+        self.headers: HeadersCollection = HeadersCollection()
+        self.request_options = []  # type: ignore
+        self.current_page = self.convert_to_page(response)
+        self.object_type = self.current_page.value[
+            0].__class__.__name__ if self.current_page.value else None
+        page = self.current_page
+
+        if page is not None:
+            self.current_page = page
+            self.has_next = bool(page.odata_next_link)
+
+    def set_headers(self, headers: dict) -> None:
+        """
+        Sets the headers for the HTTP requests.
+        This method takes a dictionary of headers and adds them to the
+        existing headers.
+        Args:
+            headers (dict): A dictionary of headers to add. The keys are the 
+            header names and the values are the header values.
+        """
+        self.headers.add_all(**headers)
+
+    def set_request_options(self, request_options: list) -> None:
+        """
+        Sets the request options for the HTTP requests.
+        Args:
+            request_options (list): The request options to set.
+        """
+        self.request_options = request_options
+
```

Comment:
I think we should also probably expose getters for
- The next_link that the iterator is currently using for the page
- The delta_link if it's returned at the end. 

In the event paging is paused/ends, these will be useful for the caller to make a request for another page using them and them pass the result object to the iterator to move forward with the other results(or store them for later use in a situation like delta requests).